### PR TITLE
feat(orc8r): adding auto-generated certs on pods.

### PR DESCRIPTION
**Pull request by ShubhamTatvamasi**
Signed-off-by: Shubham Tatvamasi <shubhamtatvamasi@gmail.com>

## Summary

Related: https://github.com/magma/magma/issues/6173

For this, to work we will have to create new Nginx image.

For testing we can use this image:
```bash
docker pull shubhamtatvamasi/nginx:magma-master-certs.0.1.1
```

## Test Plan

I have been using the code below for testing this PR.

https://github.com/ShubhamTatvamasi/magma-local

Clone the branch of PR:
```bash
git clone -b add-cert-manager-secrets-on-orc8r \
  https://github.com/ShubhamTatvamasi/magma.git --depth 1
```

https://www.youtube.com/watch?v=dXVfgJ2-5dk

## Additional Information

- [x] This change is backwards-breaking

### CODE REVIEW

**Review by hcgatewood — COMMENTED** [review]
What's the status of this PR? If it's not ready for review, please revert it to draft 🙂

**Comment by ShubhamTatvamasi** [comment]
> What's the status of this PR? If it's not ready for review, please revert it to draft

PR is ready, I just have to change the version number of helm charts. Which I thought I will be doing after a review, so I won't have to change it again in the end.
But I can do this now also.

**Comment by ShubhamTatvamasi** [comment]
So I broked something, here. Working on a fix

**Comment by ShubhamTatvamasi** [comment]
I have updated the Helm chart versions.

**Comment by ekowtaylor** [comment]
@magma/approvers-nms , @magma/approvers-orc8r , @xjtian this PR is ready for review

**Comment by ShubhamTatvamasi** [comment]
As privately discussed with @hcgatewood we will revisit this PR on August 1, 2021.

**Comment by ShubhamTatvamasi** [comment]
@hcgatewood do you think it's time we should work on this PR?

**Comment by ShubhamTatvamasi** [comment]
As per our discussion we will come back to this on September 1.

**Comment by ShubhamTatvamasi** [comment]
@hcgatewood do you think it's time we should work on this PR?

**Comment by ekowtaylor** [comment]
@hcgatewood waiting on your feedback

**Comment by hcgatewood** [comment]
@ShubhamTatvamasi I am just about ready. I need another 2 weeks or so to solidify some stuff, then I'm ready to knock this out

**Comment by ShubhamTatvamasi** [comment]
> @ShubhamTatvamasi I am just about ready. I need another 2 weeks or so to solidify some stuff, then I'm ready to knock this out

Okay, then let's get back to this on 1st October 2021.

**Comment by ShubhamTatvamasi** [comment]
@hcgatewood when can we start working on this?

File: orc8r/cloud/helm/orc8r/charts/nms/values.yaml
Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/charts/nms/values.yaml]: cc @andreilee for whether this should be Postgres at this point?

File: orc8r/cloud/helm/orc8r/charts/nms/values.yaml
Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/charts/nms/values.yaml:97]: Hm. I assumed it already was a LB type... what am I missing?

File: orc8r/cloud/helm/orc8r/values.yaml
Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/values.yaml:87]: Same as my below question... maybe this is just an incorrect default that always gets overridden? But either way, yes seems like this should indeed be denoted as LB type

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Excited for this!

One main note from offline: we want to provide 3 cert options

1. Manual. Aka the `orc8r/secrets` chart
2. Managed + self-signed. Aka what this PR will achieve
3. Managed + trusted. Via letsencrypt (only applicable for AWS deployments at this point)

The actionable for this PR is: we want to preserve support for both option (1) and option (2), where the choice is configured by an enum in the `orc8r` chart. Let me know if this makes sense!

On that note: I think the goal here is to make this PR itself still fully backward compatible. Then, if we want, in a final small PR, we can make the cut-over such that option (2) or (3) become the new default

-----

Thanks for all your great work @ShubhamTatvamasi!

**Comment by ShubhamTatvamasi** [comment]
> Excited for this!
> 
> One main note from offline: we want to provide 3 cert options
> 
> 1. Manual. Aka the `orc8r/secrets` chart
> 2. Managed + self-signed. Aka what this PR will achieve
> 3. Managed + trusted. Via letsencrypt (only applicable for AWS deployments at this point)
> 
> The actionable for this PR is: we want to preserve support for both option (1) and option (2), where the choice is configured by an enum in the `orc8r` chart. Let me know if this makes sense!
> 
> On that note: I think the goal here is to make this PR itself still fully backward compatible. Then, if we want, in a final small PR, we can make the cut-over such that option (2) or (3) become the new default
> 
> Thanks for all your great work @ShubhamTatvamasi!

Understood, I will make these changes.

File: orc8r/cloud/helm/orc8r/charts/nms/values.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/nms/values.yaml:97]: This branch has not been updated for a while so I will pull the master branch code on this.

File: orc8r/cloud/helm/orc8r/values.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/values.yaml:87]: Same https://github.com/magma/magma/pull/6413#discussion_r740730205

File: orc8r/cloud/helm/orc8r/charts/nms/values.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/nms/values.yaml]: Same https://github.com/magma/magma/pull/6413#discussion_r740730205

**Comment by ShubhamTatvamasi** [comment]
@hcgatewood right now we have hardcoded the path for the certificates on our helm charts and template files.
As our goal is to support 3 different methods for certificates. It will require 3 times the code for handling all the scenarios for Manual, self-signed and trusted certificates. Also, we will have to do changes on terraform side for supporting these changes on AWS. This will take more time than I thought.

**Comment by ekowtaylor** [comment]
What next steps are you proposing?

**Comment by ShubhamTatvamasi** [comment]
> What next steps are you proposing?

Hi @ekowtaylor, I have started coding these new helm charts and it's dependent libraries. I will update this PR after the code is ready for review. If you want we can close this PR until then.

**Comment by hcgatewood** [comment]
> @hcgatewood right now we have hardcoded the path for the certificates on our helm charts and template files. As our goal is to support 3 different methods for certificates. It will require 3 times the code for handling all the scenarios for Manual, self-signed and trusted certificates. Also, we will have to do changes on terraform side for supporting these changes on AWS. This will take more time than I thought.

Sounds good, no worries. It would be nice to modularize things as much as possible, but given your experience I'm perfectly happy to defer to your thoughts on how to structure this

**Comment by hcgatewood** [comment]
> What next steps are you proposing?

No need to close/draft the PR this point -- we're in the incorporating feedback phase

File: orc8r/cloud/helm/orc8r/charts/nms/values.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/nms/values.yaml:97]: We have set this on AWS values template file but not by default.
https://github.com/magma/magma/blob/master/orc8r/cloud/deploy/terraform/orc8r-helm-aws/templates/orc8r-values.tpl#L204

**Review by hcgatewood — APPROVED** [review]
This looks great! The one small request I have is: we're using `certs.enabled`, which feels like it could be a bit confusing since we have "certs" either way. Would it make sense to call it `managedCerts.enabled` ?

Either way, msg me when ready and I'll merge!

**Comment by ShubhamTatvamasi** [comment]
> This looks great! The one small request I have is: we're using `certs.enabled`, which feels like it could be a bit confusing since we have "certs" either way. Would it make sense to call it `managedCerts.enabled` ?
> 
> Either way, msg me when ready and I'll merge!

Good point, I was also concerned about the same issue. I am changing it to `managedCerts.enabled` for keeping it simple.

**Comment by ShubhamTatvamasi** [comment]
> > This looks great! The one small request I have is: we're using `certs.enabled`, which feels like it could be a bit confusing since we have "certs" either way. Would it make sense to call it `managedCerts.enabled` ?
> > Either way, msg me when ready and I'll merge!
> 
> Good point, I was also concerned about the same issue. I am changing it to `managedCerts.enabled` for keeping it simple.

@hcgatewood I will try to change `certs.enabled` in the upcoming PR as this change will require a lot of changes to be done on other helm charts. So for keeping it simple let's merge it as it is.

**Comment by ShubhamTatvamasi** [comment]
Our current secrets sub-chart also create configs that are stored as secrets, we will also have to move these to configmaps for fully deprecating secrets charts. I will take this up in the next PR.
```
orc8r-secrets-certs
orc8r-secrets-configs-cwf
orc8r-secrets-configs-orc8r
orc8r-secrets-envdir
orc8r-secrets-registry
```

**Comment by hcgatewood** [comment]
Looks like the deploy-sync check (cloud-test) failed -- there are 2 other places where we need to update the default helm chart values, 1 in the Terraform and 1 in the experimental "orcl" deployer

Search for `orc8r_chart_version` to find the 2 places

**Comment by ShubhamTatvamasi** [comment]
> Looks like the deploy-sync check (cloud-test) failed -- there are 2 other places where we need to update the default helm chart values, 1 in the Terraform and 1 in the experimental "orcl" deployer
> 
> Search for `orc8r_chart_version` to find the 2 places


@hcgatewood I have updated the values as requested.

**Review by themarwhal — APPROVED** [review]

**Comment by magma-alex** [comment]
Hello Shubham,

This is a very good tutorial. I'm new to Linux and am trying to replicate what you're doing to get Magma up and running. 

Would you be able to list all the dependencies that your modified Magma requires to be set up without errors? 

For instance, to run “kind create cluster”, you need to have Go and Docker installed. Is there anything else that’s needed? 

I keep getting an error saying that it refused connection to localhost:8080, then another error saying it can’t find the admin config file in the home directory. 

In other words, I’m stuck. 

Any help or guidance will be very much appreciated. 

Thanks,
Alex

**Comment by ShubhamTatvamasi** [comment]
> Hello Shubham,
> 
> This is a very good tutorial. I'm new to Linux and am trying to replicate what you're doing to get Magma up and running.
> 
> Would you be able to list all the dependencies that your modified Magma requires to be set up without errors?
> 
> For instance, to run “kind create cluster”, you need to have Go and Docker installed. Is there anything else that’s needed?
> 
> I keep getting an error saying that it refused connection to localhost:8080, then another error saying it can’t find the admin config file in the home directory.
> 
> In other words, I’m stuck.
> 
> Any help or guidance will be very much appreciated.
> 
> Thanks, Alex

For Kind setup, you can start from this video.
https://www.youtube.com/watch?v=OZw5ZJKdo44

**Pull request merged**
